src/realm/sort_descriptor.cpp

```diff
@@ -556,9 +562,18 @@ void DescriptorOrdering::append_sort(SortDescriptor sort, SortDescriptor::MergeM
         return;
     }
     if (!m_descriptors.empty()) {
-        if (SortDescriptor* previous_sort = dynamic_cast<SortDescriptor*>(m_descriptors.back().get())) {
-            previous_sort->merge(std::move(sort), mode);
-            return;
+        if (mode == SortDescriptor::MergeMode::replace) {
+            for (auto it = m_descriptors.begin(); it != m_descriptors.end(); ++it) {
```

Comment:
This loop skips checking the descriptor after each one it removes. Using `m_descriptors.erase(std::remove_if(m_descriptors.begin(), m_descriptors.end(), [](auto& desc) { return desc->get_type() == DescriptorType::Sort; }), m_descriptors.end());` rather than a manual loop would fix that and perform insignificantly better if there's more than one sort descriptor.

Reply:
Good catch.